exec/java-exec/src/main/java/org/apache/drill/exec/rpc/security/AuthenticationMechanismFactory.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.CommonConfigurationKeys;
+import org.apache.hadoop.security.UserGroupInformation;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class AuthenticationMechanismFactory implements AutoCloseable {
+  private static final org.slf4j.Logger logger =
+      org.slf4j.LoggerFactory.getLogger(AuthenticationMechanismFactory.class);
+
+  private final Map<String, AuthenticationMechanism> mechanisms = CaseInsensitiveMap.newHashMapWithExpectedSize(5);
+
+  @SuppressWarnings("unchecked")
+  public AuthenticationMechanismFactory(final ScanResult scan, final DrillConfig config,
+                                        final List<String> configuredMechanisms)
+      throws DrillbitStartupException {
+    logger.debug("Configuring authentication mechanisms: {}", configuredMechanisms);
+    // transform all names to uppercase
+    final Set<String> configuredMechanismsSet = Sets.newHashSet(Iterators.transform(configuredMechanisms.iterator(),
+        new Function<String, String>() {
+          @Nullable
+          @Override
+          public String apply(@Nullable String input) {
+            return input == null ? null : input.toUpperCase();
+          }
+        }));
+
+    // First, load Drill provided out-of-box mechanisms
+    if (configuredMechanismsSet.contains(PlainMechanism.SIMPLE_NAME)) {
+      logger.trace("Plain mechanism enabled.");
+      // instantiated here, but closed in PlainMechanism#close
+      final UserAuthenticator userAuthenticator = UserAuthenticatorFactory.createAuthenticator(config, scan);
+      mechanisms.put(PlainMechanism.SIMPLE_NAME, new PlainMechanism(userAuthenticator));
+      configuredMechanismsSet.remove(PlainMechanism.SIMPLE_NAME);
+    }
+
+    if (configuredMechanismsSet.contains(KerberosMechanism.SIMPLE_NAME)) {
+      logger.trace("Kerberos mechanism enabled.");
+      final String servicePrincipal = config.getString("drill.exec.security.auth.principal");
+      final String keytab = config.getString("drill.exec.security.auth.keytab");
+
+      try { // Kerberos mechanism requires a service to login
+        final Configuration conf = new Configuration();
+        conf.set(CommonConfigurationKeys.HADOOP_SECURITY_AUTHENTICATION, "KERBEROS");
+        // To parse non-typical principal name, uncomment below line
+        // CommonConfigurationKeys.HADOOP_SECURITY_AUTH_TO_LOCAL, rules);
+        UserGroupInformation.setConfiguration(conf);
+        UserGroupInformation.loginUserFromKeytab(servicePrincipal, keytab);
+        logger.trace("Login successful for user: {}", servicePrincipal);
+      } catch (IOException e) {
+        throw new DrillbitStartupException("Drillbit service login failed", e);
```

Comment:
Rather than failing with an exception can we simply not add Kerberos to the list of mechanisms i.e. we do not error out but rather add valid mechanisms.